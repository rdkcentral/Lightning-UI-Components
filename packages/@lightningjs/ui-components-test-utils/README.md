<!--
  Copyright 2023 Comcast Cable Communications Management, LLC

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.

  SPDX-License-Identifier: Apache-2.0
-->

# Lightning Component Test Utils

This utility has been created to assist with writing Jest tests for applications utilizing [@lightningjs/ui-components](https://www.npmjs.com/package/@lightningjs/ui-components).

## Installation

Install from NPM:

```bash
npm install --save @lightningjs/ui-components-test-utils
```

`@lightningjs/ui-components-test-utils` has a peer dependency on the [Lightning package](https://www.npmjs.com/package/@lightningjs/core).

```sh
npm install -S @lightningjs/ui @lightningjs/core
```
