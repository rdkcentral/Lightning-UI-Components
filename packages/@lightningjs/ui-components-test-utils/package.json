{
  "name": "@lightningjs/ui-components-test-utils",
  "version": "1.3.0",
  "description": "Helpful Jest unit test utilities for LightningJS applications utilizing the Lightning-UI-Components package.",
  "repository": {
    "url": "https://github.com/rdkcentral/Lightning-UI-Components",
    "directory": "packages/@lightningjs/ui-components-test-utils"
  },
  "homepage": "https://rdkcentral.github.io/Lightning-UI-Components/",
  "type": "module",
  "exports": "./index.js",
  "files": [
    "index.js",
    "src/**"
  ],
  "scripts": {
    "build": "rm -rf ./dist && rollup --config ../../../rollup/rollup.config.js --name lightning-ui-components"
  },
  "peerDependencies": {
    "@lightningjs/core": "^2.9.0"
  },
  "devDependencies": {
    "@lightningjs/ui-components-theme-base": "workspace:^",
    "jest": "^29.3.1",
    "jest-environment-jsdom": "^29.3.1",
    "jest-sinon": "^1.1.0",
    "jest-webgl-canvas-mock": "^0.2.3",
    "semantic-release": "^19.0.5",
    "semantic-release-monorepo": "^7.0.5",
    "sinon": "^15.2.0"
  },
  "dependencies": {
    "@lightningjs/ui-components": "workspace:^"
  },
  "publishConfig": {
    "access": "public"
  }
}
