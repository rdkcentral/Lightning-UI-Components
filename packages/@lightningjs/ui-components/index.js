/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

// Utils
export { default as utils } from './src/utils';

// Components
export { default as Artwork } from './src/components/Artwork/Artwork';
export { default as Badge } from './src/components/Badge/Badge';
export { default as Base } from './src/components/Base/Base';
export { default as Button } from './src/components/Button/Button';
export { default as ButtonSmall } from './src/components/Button/ButtonSmall';
export {
  default as Card,
  CardRadio,
  CardSection,
  CardTitle
} from './src/components/Card';
export {
  default as CardContent,
  CardContentHorizontal,
  CardContentHorizontalLarge,
  CardContentVertical,
  CardContentVerticalSmall
} from './src/components/CardContent';
export { default as Checkbox } from './src/components/Checkbox/Checkbox';
export { default as CheckboxSmall } from './src/components/Checkbox/CheckboxSmall';
export { default as Column } from './src/components/Column/Column';
export { default as Control } from './src/components/Control/Control';
export { default as ControlSmall } from './src/components/Control/ControlSmall';
export { default as ControlRow } from './src/components/ControlRow/ControlRow';
export { default as FocusManager } from './src/components/FocusManager/FocusManager';
export { default as Gradient } from './src/components/Gradient/Gradient';
export { default as GridOverlay } from './src/components/GridOverlay/GridOverlay';
export { default as Icon } from './src/components/Icon/Icon';
export { default as InlineContent } from './src/components/InlineContent/InlineContent';
export { default as Input } from './src/components/Input/Input';
export { default as Key } from './src/components/Key/Key';
export {
  default as Keyboard,
  KeyboardEmail,
  KeyboardFullscreen,
  KeyboardNumbers,
  KeyboardQwerty,
  KeyboardInput,
  KeyboardSearch
} from './src/components/Keyboard';
export { default as Knob } from './src/components/Knob/Knob';
export { default as Label } from './src/components/Label/Label';
export {
  default as ListItem,
  ListItemPicker,
  ListItemSlider
} from './src/components/ListItem';
export { default as Marquee } from './src/components/Marquee/Marquee';
export { default as MetadataBase } from './src/components/MetadataBase/MetadataBase';
export { default as MetadataCard } from './src/components/MetadataCard';
export { default as MetadataCardContent } from './src/components/MetadataCardContent';
export { default as MetadataTile } from './src/components/MetadataTile/MetadataTile';
export { default as NavigationManager } from './src/components/NavigationManager';
export { default as ProgressBar } from './src/components/ProgressBar/ProgressBar';
export { default as Provider } from './src/components/Provider';
export { default as Radio } from './src/components/Radio/Radio';
export { default as RadioSmall } from './src/components/Radio/RadioSmall';
export { default as Row } from './src/components/Row/Row';
export { default as ScrollWrapper } from './src/components/ScrollWrapper/ScrollWrapper';
export { default as Shadow } from './src/components/Shadow/Shadow';
export { default as Slider } from './src/components/Slider/Slider';
export { default as SliderLarge } from './src/components/Slider/SliderLarge';
export { default as Surface } from './src/components/Surface/Surface';
export { default as TabBar, Tab } from './src/components/TabBar';
export { default as TextBox } from './src/components/TextBox/TextBox';
export { default as Tile } from './src/components/Tile/Tile';
export { default as TitleRow } from './src/components/TitleRow/TitleRow';
export { default as Toggle } from './src/components/Toggle/Toggle';
export { default as ToggleSmall } from './src/components/Toggle/ToggleSmall';
export { default as Tooltip } from './src/components/Tooltip/Tooltip';

// Globals
export { default as context } from './src/globals/context';
export { updateManager } from './src/globals/global-update-manager/GlobalUpdateManager';
export { default as pool } from './src/globals/pool';

// Mixins
export {
  default as withAnnouncer,
  generateAbbrevConfig,
  defaultAbbrevConfig
} from './src/mixins/withAnnouncer';
export { default as Speech } from './src/mixins/withAnnouncer/Speech';
export { default as withClassCache } from './src/mixins/withClassCache';
export { default as withHandleKey } from './src/mixins/withHandleKey';
export { default as withLayout } from './src/mixins/withLayout';
export { default as withMarqueeSync } from './src/mixins/withMarqueeSync';
export { default as withSelections } from './src/mixins/withSelections';
export { default as withThemeStyles } from './src/mixins/withThemeStyles';
export { default as withTags } from './src/mixins/withTags';
export { default as withUpdates } from './src/mixins/withUpdates';
export { default as withEditItems } from './src/mixins/withEditItems';
export { default as withExtensions } from './src/mixins/withExtensions';

// Shaders
export { default as FadeShader } from './src/shaders/FadeShader';

// Textures
export { default as LinearGradient } from './src/textures/LinearGradient';
export { default as Circle } from './src/textures/Circle';
export { default as Arrow } from './src/textures/Arrow';
export { default as Bubble } from './src/textures/Bubble';
export { default as Line } from './src/textures/Line';

// Storybook
export { default as SignalButton } from './src/docs/story-components/SignalButton';
