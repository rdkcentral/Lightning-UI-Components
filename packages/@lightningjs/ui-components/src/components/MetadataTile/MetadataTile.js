/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import MetadataBase from '../MetadataBase';
import * as styles from './MetadataTile.styles.js';

export default class MetadataTile extends MetadataBase {
  static get __componentName() {
    return 'MetadataTile';
  }

  static get __themeStyle() {
    return styles;
  }

  _updateSubtitle() {
    if (this.description) {
      this._Subtitle.patch({ content: '' });
      this._Subtitle.alpha = 0;
      this._Subtitle.visible = false;
    } else {
      this._Subtitle.alpha = 1;
      this._Subtitle.visible = true;
      super._updateSubtitle();
    }
  }

  _updateSubtitleLayout({ h }) {
    if (this.subtitle && !this.description) {
      this._SubtitleWrapper.h = h;
      this._SubtitleWrapper.alpha = this.style.alpha;
    } else {
      this._SubtitleWrapper.h = 0;
    }
  }

  set announce(announce) {
    super.announce = announce;
  }

  get announce() {
    return (
      this._announce || [
        this._Title && this._Title.announce,
        (this._Subtitle && this._Subtitle.announce) ||
          (this._Description && this._Description.announce),
        this.logoTitle
      ]
    );
  }

  _textH() {
    const titleH = (this.title && this._Title && this._Title.h) || 0;
    const subtitleH =
      (this.subtitle &&
        this._Subtitle &&
        this._Subtitle.visible &&
        this._SubtitleWrapper.h) ||
      0;
    const descriptionH =
      (this.description && this._Description && this._Description.h) || 0;
    return titleH + subtitleH + descriptionH;
  }
}
