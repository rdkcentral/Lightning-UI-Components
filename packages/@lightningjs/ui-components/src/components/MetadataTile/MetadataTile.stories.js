/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import lng from '@lightningjs/core';
import lightningbolt from '../../assets/images/ic_lightning_white_32.png';
import circle from '../../assets/images/circle.svg';
import { getHexColor } from '../../utils';
import { createModeControl } from '../../docs/utils';
import { default as MetadataTileComponent } from '.';
import mdx from './MetadataTile.mdx';
import { CATEGORIES } from '../../docs/constants';

export default {
  title: 'Metadata/MetadataTile',
  parameters: {
    docs: {
      page: mdx
    }
  }
};

export const MetadataTile = args =>
  class MetadataTile extends lng.Component {
    static _template() {
      return {
        MetadataTile: {
          type: MetadataTileComponent,
          w: args.w,
          title: args.title,
          subtitle: args.subtitle,
          description: args.description,
          logo: args.logo !== 'none' ? circle : null,
          logoTitle: args.logo !== 'none' ? args.logoTitle : null,
          logoPosition: args.logoPosition
        }
      };
    }

    _getFocused() {
      return this.tag('MetadataTile');
    }
  };

MetadataTile.storyName = 'MetadataTile';

MetadataTile.args = {
  w: 400,
  title: 'Title',
  subtitle: [
    '94%',
    {
      icon: lightningbolt,
      style: { color: getHexColor('00ff00') },
      title: 'Green Lightning Bolt'
    },
    '86%',
    {
      icon: 'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
      title: 'Rotten Tomatoes rating'
    }
  ],
  description: 'Description',
  logo: circle,
  logoTitle: 'Logo title',
  logoPosition: 'right',
  mode: 'unfocused'
};

MetadataTile.argTypes = {
  ...createModeControl({ summaryValue: MetadataTile.args.mode }),
  w: {
    control: 'number',
    description: 'Width of component',
    table: {
      defaultValue: { summary: 0 }
    }
  },
  title: {
    control: 'text',
    description: 'Title text',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  subtitle: {
    control: 'object',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  description: {
    control: 'text',
    description: 'description text',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  logo: {
    control: 'select',
    options: ['none', circle],
    description: 'Image to use for logo',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  logoTitle: {
    control: 'text',
    description: 'Title to use for logo in announcer',
    table: {
      defaultValue: { summary: 'undefined' }
    }
  },
  logoPosition: {
    control: 'select',
    options: ['right', 'left'],
    description: 'Position logo on the left/right side',
    table: {
      defaultValue: { summary: 'right' }
    }
  }
};
