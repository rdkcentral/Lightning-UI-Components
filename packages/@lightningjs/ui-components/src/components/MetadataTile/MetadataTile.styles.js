/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

export const base = theme => ({
  titleTextStyle: theme.typography.headline3,
  descriptionTextStyle: theme.typography.body3
});

export const tone = theme => ({
  neutral: {
    detailsTextStyle: { textColor: theme.color.textNeutralSecondary },
    descriptionTextStyle: { textColor: theme.color.textNeutral },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled },
        descriptionTextStyle: {
          textColor: theme.color.textNeutralDisabled
        }
      }
    }
  },
  inverse: {
    detailsTextStyle: { textColor: theme.color.textInverseSecondary },
    descriptionTextStyle: { textColor: theme.color.textInverse },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled },
        descriptionTextStyle: {
          textColor: theme.color.textNeutralDisabled
        }
      }
    }
  },
  brand: {
    detailsTextStyle: { textColor: theme.color.textNeutralSecondary },
    descriptionTextStyle: { textColor: theme.color.textNeutral },
    mode: {
      disabled: {
        detailsTextStyle: { textColor: theme.color.textNeutralDisabled },
        descriptionTextStyle: {
          textColor: theme.color.textNeutralDisabled
        }
      }
    }
  }
});
