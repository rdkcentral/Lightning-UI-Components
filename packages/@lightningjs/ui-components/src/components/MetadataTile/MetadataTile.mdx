import { Meta } from '@storybook/blocks';
import * as MetadataTileStories from './MetadataTile.stories.js';

<Meta of={MetadataTileStories} />

# MetadataTile

Metadata to be displayed along with `Tiles`.

## Source

https://github.com/rdkcentral/Lightning-UI-Components/blob/develop/packages/@lightningjs/ui-components/src/components/MetadataTile/MetadataTile.js

## Usage

MetadataTile can be used with title (text), details (inline content) or description (text), and/or a logo.

```js
import { MetadataTile } from '@lightningjs/ui-components';
import lightningbolt from '../../assets/images/ic_lightning_white_32.png';
import circle from '../../assets/images/circle.svg';

class Example extends lng.Component {
  static _template() {
    return {
      MetadataTile: {
        type: MetadataTile,
        w: 400,
        h: 300,
        title: 'Title',
        details: [
          '94%',
          {
            icon: lightningbolt,
            style: { color: getHexColor('00ff00') },
            title: 'Green Lightning Bolt'
          },
          '86%',
          {
            icon: 'http://myriad.merlin.comcast.com/select/logo?entityId=8527084350383982239&width=32&height=&ratio=1x1&trim=false',
            title: 'Rotten Tomatoes rating'
          }
        ],
        logo: circle,
        logoWidth: 32,
        logoHeight: 32
      }
    };
  }
}
```

## API

### Properties

| name         | type           | required             | default   | description                                                             |
| ------------ | -------------- | -------------------- | --------- | ----------------------------------------------------------------------- |
| description  | string         | false                | undefined | description text to place under title (hides details)                   |
| details      | inline content | false                | undefined | relevant content data in the middle (hidden if description is provided) |
| logo         | string         | false                | undefined | logo to display at bottom of component                                  |
| logoHeight   | number         | true (if using logo) | undefined | height of logo                                                          |
| logoPosition | string         | false                | 'right'   | which side to place logo (`right` or `left`)                            |
| logoTitle    | string         | true (if using logo) | undefined | title of logo to use for announcer                                      |
| logoWidth    | number         | true (if using logo) | undefined | width of logo                                                           |
| title        | string         | false                | undefined | first line or headline of the content                                   |

### Style Properties

| name                 | type             | description                                                     |
| -------------------- | ---------------- | --------------------------------------------------------------- |
| descriptionTextStyle | string \| object | text style for description                                      |
| fadeWidth            | number           | width of fade applied to second line if text encroaches on logo |
| logoHeight           | number           | height for logo                                                 |
| logoPadding          | number           | spacing between logo and secondLine                             |
| logoWidth            | number           | width for logo                                                  |
| detailsTextStyle     | string \| object | text style for details                                          |
| titleTextStyle       | string \| object | text style for title                                            |
