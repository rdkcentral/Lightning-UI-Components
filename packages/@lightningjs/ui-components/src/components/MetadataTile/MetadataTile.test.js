/**
 * Copyright 2023 Comcast Cable Communications Management, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { makeCreateComponent } from '@lightningjs/ui-components-test-utils';
import MetadataTile from './index.js';

const createComponent = makeCreateComponent(MetadataTile, {
  w: 400,
  h: 300
});

describe('MetadataTile', () => {
  let metadataTile, testRenderer;

  beforeEach(() => {
    [metadataTile, testRenderer] = createComponent(
      {},
      {
        spyOnMethods: ['_update', '_subtitleLoaded']
      }
    );
  });
  afterEach(() => {
    metadataTile = null;
    testRenderer = null;
  });

  it('renders', () => {
    const tree = testRenderer.toJSON(2);
    expect(tree).toMatchSnapshot();
  });

  it('sets the announce string to the appropriate text content status', () => {
    const title = 'Title';
    const subtitle = 'Subtitle';
    const description = 'Description';
    const logoTitle = 'Logo Title';
    metadataTile.patch({ title, subtitle, logoTitle });
    testRenderer.forceAllUpdates();
    expect(metadataTile.announce).toEqual([title, subtitle, logoTitle]);
    metadataTile.patch({ title, subtitle: undefined, description, logoTitle });
    testRenderer.forceAllUpdates();
    expect(metadataTile.announce).toEqual([title, description, logoTitle]);
  });

  it('overrides the announce string', () => {
    const overrideString = 'Custom announce string';
    metadataTile.announce = overrideString;
    testRenderer.forceAllUpdates();
    expect(metadataTile.announce).toBe(overrideString);
  });

  it('hides subtitle if description is provided', async () => {
    metadataTile.subtitle = 'subtitle';
    await metadataTile.__updateSpyPromise;
    expect(metadataTile._Subtitle.alpha).toBe(1);
    expect(metadataTile._Subtitle.visible).toBe(true);
    metadataTile.description = 'description';
    await metadataTile.__updateSpyPromise;
    expect(metadataTile._Subtitle.alpha).toBe(0);
    expect(metadataTile._Subtitle.visible).toBe(false);
    expect(metadataTile._SubtitleWrapper.h).toBe(0);
  });

  it('should render a subtitle', async () => {
    const subtitle = 'subtitle text';
    expect(metadataTile._SubtitleWrapper.h).toBe(0);
    metadataTile.subtitle = subtitle;
    await metadataTile.__subtitleLoadedSpyPromise;
    testRenderer.forceAllUpdates();
    expect(metadataTile._Subtitle.content).toBe(subtitle);
    expect(metadataTile._SubtitleWrapper.h).toBeGreaterThan(0);
    expect(metadataTile._SubtitleWrapper.alpha).toBe(1);
  });
});
