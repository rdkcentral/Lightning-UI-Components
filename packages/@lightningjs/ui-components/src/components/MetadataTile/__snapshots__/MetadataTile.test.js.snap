// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MetadataTile renders 1`] = `
{
  "Component": {
    "Logo": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "clipping": false,
      "color": 4294506490,
      "enabled": true,
      "flex": false,
      "flexItem": false,
      "h": 32,
      "hasFinalFocus": false,
      "hasFocus": false,
      "isComponent": true,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Logo",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": "",
      "tag": [Function],
      "tags": [
        "Logo",
      ],
      "type": "Icon",
      "visible": true,
      "w": 32,
      "x": 368,
      "y": 0,
      "zIndex": 0,
    },
    "Text": {
      "active": false,
      "alpha": 1,
      "attached": true,
      "boundsMargin": null,
      "children": {
        "Description": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Description",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Description",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "SubtitleWrapper": {
          "active": false,
          "alpha": 1,
          "attached": true,
          "boundsMargin": null,
          "children": {
            "Subtitle": {
              "active": false,
              "alpha": 1,
              "attached": true,
              "boundsMargin": null,
              "clipping": false,
              "color": 4294967295,
              "enabled": false,
              "flex": false,
              "flexItem": false,
              "h": 0,
              "hasFinalFocus": false,
              "hasFocus": false,
              "isComponent": true,
              "mount": 0,
              "mountX": 0,
              "mountY": 0,
              "pivot": 0.5,
              "pivotX": 0.5,
              "pivotY": 0.5,
              "ref": "Subtitle",
              "renderOfScreen": undefined,
              "renderToTexture": false,
              "scale": 1,
              "scaleX": 1,
              "scaleY": 1,
              "state": "",
              "tag": [Function],
              "tags": [
                "Subtitle",
              ],
              "type": "TextBox",
              "visible": true,
              "w": 0,
              "x": 0,
              "y": 0,
              "zIndex": 0,
            },
          },
          "clipping": false,
          "color": 4294967295,
          "enabled": false,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "isComponent": undefined,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "SubtitleWrapper",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": undefined,
          "tag": [Function],
          "tags": [
            "SubtitleWrapper",
          ],
          "visible": false,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
        "Title": {
          "active": false,
          "alpha": 0.001,
          "attached": true,
          "boundsMargin": null,
          "clipping": false,
          "color": 4294967295,
          "enabled": true,
          "flex": false,
          "flexItem": true,
          "h": 0,
          "hasFinalFocus": false,
          "hasFocus": false,
          "isComponent": true,
          "mount": 0,
          "mountX": 0,
          "mountY": 0,
          "pivot": 0.5,
          "pivotX": 0.5,
          "pivotY": 0.5,
          "ref": "Title",
          "renderOfScreen": undefined,
          "renderToTexture": false,
          "scale": 1,
          "scaleX": 1,
          "scaleY": 1,
          "state": "",
          "tag": [Function],
          "tags": [
            "Title",
          ],
          "type": "TextBox",
          "visible": true,
          "w": 0,
          "x": 0,
          "y": 0,
          "zIndex": 0,
        },
      },
      "clipping": false,
      "color": 4294967295,
      "enabled": true,
      "flex": true,
      "flexItem": true,
      "h": 0,
      "isComponent": undefined,
      "mount": 0,
      "mountX": 0,
      "mountY": 0,
      "pivot": 0.5,
      "pivotX": 0.5,
      "pivotY": 0.5,
      "ref": "Text",
      "renderOfScreen": undefined,
      "renderToTexture": false,
      "scale": 1,
      "scaleX": 1,
      "scaleY": 1,
      "state": undefined,
      "tag": [Function],
      "tags": [
        "Text",
      ],
      "visible": true,
      "w": 400,
      "x": 0,
      "y": 16,
      "zIndex": 0,
    },
  },
}
`;
